### PR TITLE
add quic e2e test

### PR DESCRIPTION
**Pull request by wackxu**
**What type of PR is this?**

/kind test

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #

add quic e2e test for CI

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

File: cloud/pkg/edgecontroller/controller/upstream.go [hunk 1]
```diff
@@ -872,15 +871,9 @@ func (uc *UpstreamController) patchNode() {
...
 			}
 
-			dataReq, ok := msg.Content.(string)
-			if !ok {
-				klog.Warningf("message: %s process failure, get data failed with error: %v", msg.GetID(), err)
-				continue
-			}
-
-			nodeByte, err := base64.URLEncoding.DecodeString(dataReq)
+			nodeByte, err := msg.GetContentData()
```
[the remark this answers is not in the document]
Comment (wackxu) [reply cloud/pkg/edgecontroller/controller/upstream.go:874]: yes

File: staging/src/github.com/kubeedge/viaduct/pkg/conn/quic.go [hunk 2]
```diff
@@ -297,6 +298,9 @@ func (conn *QuicConnection) WriteMessageAsync(msg *model.Message) error {
 		return fmt.Errorf("bad connection session")
 	}
 
+	conn.locker.Lock()
+	defer conn.locker.Unlock()
```
[the remark this answers is not in the document]
Comment (wackxu) [reply staging/src/github.com/kubeedge/viaduct/pkg/conn/quic.go:302]: the websocket or quic connections support one concurrent reader and one concurrent writer. use lock to protect the connections

**Review by gy95 — COMMENTED** [review]
lgtm

File: .github/workflows/main.yaml [hunk 3]
```diff
@@ -179,6 +179,56 @@ jobs:
             /tmp/cloudcore.log
             /tmp/edgecore.log
 
+  quic_e2e_test:
```
Comment (vincentgoat) [inline .github/workflows/main.yaml:182]: I think we can reuse job `e2e_test` to run `quic-e2e-test` through the github action feature:
````    
strategy:
      matrix:
````
One example can be retrieved from release.yaml.

File: .github/workflows/main.yaml [hunk 4]
```diff
@@ -169,7 +174,9 @@ jobs:
         run: |
           docker load < /home/runner/build-tools/build-tools.tar
 
-      - run: make e2e
+      - run: |
+          export ENABLE_QUIC=${{ matrix.ENABLE_QUIC }}
+          make e2e
```
Comment (gy95) [inline .github/workflows/main.yaml:179]: maybe influence logs uploaded. both two e2e will use the same logs name: in line184

File: .github/workflows/main.yaml [hunk 5]
```diff
@@ -132,6 +132,11 @@ jobs:
 
   e2e_test:
     runs-on: ubuntu-18.04
+    strategy:
+      matrix:
+        ENABLE_QUIC:
+          - false
+          - true
```
Comment (fisherxu) [inline .github/workflows/main.yaml]:
```suggestion
        protocol:
          - websocket
          - quic
```

File: hack/local-up-kubeedge.sh [hunk 6]
```diff
@@ -19,6 +19,7 @@ ENABLE_DAEMON=${ENABLE_DAEMON:-false}
 LOG_DIR=${LOG_DIR:-"/tmp"}
 LOG_LEVEL=${LOG_LEVEL:-2}
 TIMEOUT=${TIMEOUT:-60}s
+PROTOCOL=${PROTOCOL:-"WebSocket"}
```
Comment (fisherxu) [inline hack/local-up-kubeedge.sh:22]:
```suggestion
# protocol for cloud-edge communication, i.e. websocket or quic
PROTOCOL=${PROTOCOL:-websocket}
```

File: .github/workflows/main.yaml [hunk 7]
```diff
@@ -132,6 +132,11 @@ jobs:
 
   e2e_test:
     runs-on: ubuntu-18.04
+    strategy:
+      matrix:
+        PROTOCOL:
+          - WebSocket
+          - QUIC
```
Comment (fisherxu) [inline .github/workflows/main.yaml:139]:
```suggestion
          - websocket
          - quic
```
File name and shell env value look like all lowcase.

File: .github/workflows/main.yaml (hunk 5 above)
Comment (wackxu) [reply .github/workflows/main.yaml]: talked with @fisherxu ,  let us keep this

File: .github/workflows/main.yaml (hunk 7 above)
Comment (wackxu) [reply .github/workflows/main.yaml:139]: talked with @fisherxu ,  let us keep this

**Review by fisherxu — COMMENTED** [review]
/lgtm
/approve

**Pull request merged**